### PR TITLE
Remove schema validation from cross-chain/command verification in module LIPs

### LINKED ISSUE #435: Remove explicit schema validation from commands/cross-chain commands verification in module LIPs
## Motivation

Align with the implementation https://github.com/LiskHQ/lisk-sdk/issues/8475

## Proposed Specifications

Remove the explicit schema validation and add a sentence to each module LIP as similarly done in https://github.com/LiskHQ/lips/pull/419

## Affected LIPs

Module LIPs

### PR DESCRIPTION
**Pull request by AnanyaSshri**
Resolves #435

### CODE REVIEW

File: proposals/lip-0057.md [hunk 1]
```diff
@@ -870,7 +869,7 @@ unlockTransactionParams = {
 ```python
 def verify(trs: Transaction) -> None:
     trsParams = decode(unlockTransactionParams, trs.params)
```
Comment (ricott1) [inline proposals/lip-0057.md:871]: I think we can remove this line as well

File: proposals/lip-0070.md [hunk 2]
```diff
@@ -198,7 +197,7 @@ claimRewardsParams = {
 ```python
 def verify(trs: Transaction) -> None:
     trsParams = decode(claimRewardsParams, trs.params)
```
Comment (ricott1) [inline proposals/lip-0070.md:199]: We can remove this line as well

File: proposals/lip-0049.md [hunk 3]
```diff
@@ -273,6 +273,9 @@ def apply(trs: Transaction, ccm: CCM) -> None:
     else:
         crossChainCommand = cross-chain command associated with (ccm.module, ccm.crossChainCommand)
         try:  
+            #Decode and validate the parameterSchema associated with ccm module or crossChainCommand
+            ccmParams = decode(paramsSchema, ccm.params)
```
Comment (ricott1) [inline proposals/lip-0049.md:277]: `paramsSchema` is undefined. I would define it in words, similar to what we do for `crossChainCommand`

File: proposals/lip-0049.md [hunk 4]
```diff
@@ -273,6 +273,9 @@ def apply(trs: Transaction, ccm: CCM) -> None:
...
+            validateObjectSchema(channelTerminatedCCMParamsSchema, ccmParams)
```
Comment (ricott1) [inline proposals/lip-0049.md:278]:
```suggestion
            validateObjectSchema(paramsSchema, ccmParams)
```

File: proposals/lip-0049.md [hunk 5]
```diff
@@ -273,6 +273,9 @@ def apply(trs: Transaction, ccm: CCM) -> None:
     else:
         crossChainCommand = cross-chain command associated with (ccm.module, ccm.crossChainCommand)
         try:  
+            #Decode and validate the parameterSchema associated with ccm module or crossChainCommand
```
Comment (ricott1) [inline proposals/lip-0049.md:276]:
```suggestion
            #Decode and validate the parameter schema associated with ccm module and crossChainCommand
```

Comment (ricott1) [reply proposals/lip-0049.md:276]: Also, if you define the params schema as I suggest, this comment is probably superfluous

File: proposals/lip-0053.md [hunk 6]
```diff
@@ -92,7 +92,7 @@ These points guarantee that the CCMs are always forwarded to the correct chains
 The [execution of cross-chain messages][lip-0049#apply] as part of the processing of a cross-chain update is done in several stages, in a way analogous to the execution of a transaction as part of the [processing of a block][lip-0055#block-processing-stages]:
 
 1. **Cross-chain message verification**: Each module can define protocol logic that verifies a CCM, possibly by accessing the state store. If an error occurs, the channel with the CCU sending chain is terminated and the CCU execution stops.
-2. **Cross-chain command verification**: The cross-chain command corresponding to the CCM `module`-`crossChainCommand` combination is verified. If an error occurs, the channel with the CCM sending chain is terminated and the CCM execution stops.
+2. **Cross-chain command verification**: The cross-chain command corresponding to the CCM `module`-`crossChainCommand` combination is verified. If an error occurs, the channel with the CCM sending chain is terminated and the CCM execution stops. Note that before verifying the cross-chain command, the parameters are decoded and validated against the schema.
```
Comment (ricott1) [inline proposals/lip-0053.md:95]: I think we should say this at the beginning, so that it is clear that also in this case if an error occurs the channel is terminated etc

File: proposals/lip-0055.md [hunk 7]
```diff
@@ -122,7 +122,7 @@ The full processing of a block is organized as follows.
 5. **Block forwarding**: After the initial checks, the full block is forwarded to a subset of peers.
 6. **Before transactions execution**: Each module can define protocol logic that is executed *before* the transactions contained in the block are processed.
 7. **Transaction verification**: Each module can define protocol logic that verifies a transaction, possibly by accessing the state store. If an error occurs, the transaction is invalid and the whole block is discarded.
-8. **Command verification**: The command corresponding to the `module`/`command` combination is verified. If an error occurs, the transaction is invalid and the whole block is discarded.
+8. **Command verification**: The command corresponding to the `module`/`command` combination is verified. If an error occurs, the transaction is invalid and the whole block is discarded. Note that before verifying the command, the parameters are decoded and validated against the schema.
```
Comment (ricott1) [inline proposals/lip-0055.md:125]: Same as my comment above, let's say this first

**Review by ricott1 — CHANGES_REQUESTED** [review]
Some minor comments plus restructuring the explanations in LIP0055. In LIP0055 I would also be precise and explain which function is called.

File: proposals/lip-0057.md (hunk 1 above)
Comment (AnanyaSshri) [reply proposals/lip-0057.md:871]: So this will be an empty function with just a `return` statement? Shouldn't we, then, just remove the whole function?

Comment (AnanyaSshri) [reply proposals/lip-0057.md:871]: Done

File: proposals/lip-0070.md (hunk 2 above)
Comment (AnanyaSshri) [reply proposals/lip-0070.md:199]: Done

File: proposals/lip-0070.md [hunk 8]
```diff
@@ -197,8 +196,7 @@ claimRewardsParams = {
 
 ```python
 def verify(trs: Transaction) -> None:
-    trsParams = decode(claimRewardsParams, trs.params)
-    validateObjectSchema(claimRewardsParams, trsParams)
+    return
 ```
 
 #### Execution
```
Comment (gkoumout) [inline proposals/lip-0070.md:202]:
```suggestion
##### Execution
```

This typo carries over from previous versions.

File: proposals/lip-0070.md [hunk 9]
```diff
@@ -197,8 +196,7 @@ claimRewardsParams = {
 
 ```python
```
Comment (gkoumout) [inline proposals/lip-0070.md:197]: Similar to my comment for the unlock command in LIP0057, I would completely remove pseudocode and say in words that we don't perform any additional verification.

File: proposals/lip-0047.md [hunk 10]
```diff
@@ -380,7 +379,6 @@ A transaction with module name `MODULE_NAME_POA` and command name `COMMAND_UPDAT
 ```python
 def verify(trs: Transaction) -> None:
     trsParams = decode(updateGeneratorKeyParamsSchema, trs.params)
```
Comment (gkoumout) [inline proposals/lip-0047.md:381]: Here I think we could even remove this line, since `trsParams` is not used in the function.

File: proposals/lip-0049.md [hunk 11]
```diff
@@ -442,8 +445,7 @@ A channel terminated message is created by the Interoperability module when [ter
```
Comment (gkoumout) [inline proposals/lip-0049.md:445]: We can completely skip the pseudocode here.

File: proposals/lip-0049.md [hunk 12]
```diff
@@ -611,7 +612,6 @@ A sidechain terminated message is created by the Interoperability module when a
 ```python
 def verify(trs: Transaction, ccm: CCM) -> None:
     ccmParams = decode(sidechainTerminatedCCMParamsSchema, ccm.params)
```
Comment (gkoumout) [inline proposals/lip-0049.md:614]: This line could also be removed IMO.

**Review by gkoumout — CHANGES_REQUESTED** [review]
I left a first review. I still have some questions but we can discuss them offline before next review round. 

Also LIP 74 (Governance) is not using trsParams consistenly with other LIPs.

File: proposals/lip-0055.md [hunk 13]
```diff
@@ -97,7 +97,7 @@ For the rest of this proposal we define the following constants.
 | `MAX_MODULE_NAME_LENGTH` | integer | 32 | The maximum length of a string specifying the name of a module. |
 | `OWN_CHAIN_ID` | `bytes` | set in configuration | The chain ID of the chain under consideration as defined in [LIP 0037][lip-0037#own-chain-id]. |
 
-Furthermore, in the following we indicate with `block` be the block under consideration and with `previousBlock` the previous block of the chain. Calling a function `fct` from another module `module` is represented by `module.fct`.
+Furthermore, in the following we indicate with `block` be the block under consideration and with `previousBlock` the previous block of the chain. Calling a function `fct` from another module `module` is represented by `module.fct`. The functions `decode` and `validateObjectSchema` are defined in [LIP 0027][https://github.com/LiskHQ/lips/blob/main/proposals/lip-0027.md#notation-for-encoding-and-decoding].
```
Comment (ricott1) [inline proposals/lip-0055.md:100]: I would prefer to define a new link like the others, by adding it to the bottom of the LIP.

File: proposals/lip-0055.md [hunk 14]
```diff
@@ -150,7 +150,7 @@ The full generation of a block is organized as follows.
...
 After this stage has been completed, transactions are selected one-by-one from a transaction pool. The transaction processing stages (stages 4 to 8) are repeated for each transaction selected. If stages 4, 5, 6, and 8 are executed successfully, the transaction is valid and it is included in the block, otherwise it is invalid and therefore discarded.
 5. **Transaction verification**: Each module can define protocol logic that verifies a transaction, possibly by accessing the state store. If an error occurs, the transaction is invalid and it is not included in the block.
-6. **Command verification**: The command corresponding to the `module`/`command` combination is verified. If an error occurs, the transaction is invalid and it is not included in the block.
+6. **Command verification**: Before verifying the cross-chain command, the parameters are decoded using `decode` function and validated against the schema using `validateObjectSchema`. Then, the command corresponding to the `module`/`command` combination is verified. If an error occurs either during parameter validation or command verification, the transaction is invalid and it is not included in the block.
```
Comment (ricott1) [inline proposals/lip-0055.md:153]:
```suggestion
6. **Command verification**: Before verifying the command, the parameters are decoded using `decode` function and validated against the schema using `validateObjectSchema`. Then, the command corresponding to the `module`/`command` combination is verified. If an error occurs either during parameter validation or command verification, the transaction is invalid and it is not included in the block.
```

File: proposals/lip-0055.md [hunk 15]
```diff
@@ -122,7 +122,7 @@ The full processing of a block is organized as follows.
 5. **Block forwarding**: After the initial checks, the full block is forwarded to a subset of peers.
 6. **Before transactions execution**: Each module can define protocol logic that is executed *before* the transactions contained in the block are processed.
 7. **Transaction verification**: Each module can define protocol logic that verifies a transaction, possibly by accessing the state store. If an error occurs, the transaction is invalid and the whole block is discarded.
-8. **Command verification**: The command corresponding to the `module`/`command` combination is verified. If an error occurs, the transaction is invalid and the whole block is discarded.
+8. **Command verification**: Before verifying the command, the parameters are decoded using `decode` function and validated against the schema using `validateObjectSchema`. Then, the command corresponding to the `module`/`command` combination is verified. If an error occurs either during parameter validation or command verification, the transaction is invalid and the whole block is discarded.
```
Comment (ricott1) [inline proposals/lip-0055.md:125]:
```suggestion
8. **Command verification**: Before verifying the command, the parameters are decoded using `decode` function and validated against the schema using `validateObjectSchema`. Then, the command corresponding to the `module`/`command` combination is verified. If an error occurs either during parameter validation or command verification, the transaction is invalid and the whole block is discarded.
```

File: proposals/lip-0053.md [hunk 16]
```diff
@@ -92,7 +92,7 @@ These points guarantee that the CCMs are always forwarded to the correct chains
...
 
 1. **Cross-chain message verification**: Each module can define protocol logic that verifies a CCM, possibly by accessing the state store. If an error occurs, the channel with the CCU sending chain is terminated and the CCU execution stops.
-2. **Cross-chain command verification**: The cross-chain command corresponding to the CCM `module`-`crossChainCommand` combination is verified. If an error occurs, the channel with the CCM sending chain is terminated and the CCM execution stops.
+2. **Cross-chain command verification**: Before verifying the cross-chain command, the parameters are decoded using `decode` function and validated against the schema using `validateObjectSchema`. Then, the cross-chain command corresponding to the CCM `module`-`crossChainCommand` combination is verified. If an error occurs either during parameter validation or cross-chain command verification, the channel with the CCM sending chain is terminated and the CCM execution stops.
```
Comment (ricott1) [inline proposals/lip-0053.md:95]:
```suggestion
2. **Cross-chain command verification**: Before verifying the cross-chain command, the parameters are decoded using `decode` function and validated against the schema using `validateObjectSchema`. Then, the cross-chain command corresponding to the CCM `module`-`crossChainCommand` combination is verified. If an error occurs either during parameter validation or cross-chain command verification, the channel with the CCM sending chain is terminated and the CCM execution stops.
```

**Review by ricott1 — CHANGES_REQUESTED** [review]
A couple minor comments

**Review by ricott1 — APPROVED** [review]

File: proposals/lip-0049.md [hunk 17]
```diff
@@ -273,6 +273,9 @@ def apply(trs: Transaction, ccm: CCM) -> None:
     else:
         crossChainCommand = cross-chain command associated with (ccm.module, ccm.crossChainCommand)
         try:  
+            paramsSchema = parameter schema associated with (ccm.module, ccm.crossChainCommand)
```
Comment (gkoumout) [inline proposals/lip-0049.md:276]: Minor comment: Perhaps we can add comments here (before line 276 and 279) to separate the steps of validating params and verifying the command, to make it easier to parse.

File: proposals/lip-0053.md [hunk 18]
```diff
@@ -92,7 +92,7 @@ These points guarantee that the CCMs are always forwarded to the correct chains
...
```
Comment (gkoumout) [inline proposals/lip-0053.md:95]:
```suggestion
2. **Cross-chain command verification**: The parameters are decoded using `decode` function and validated against the schema using `validateObjectSchema`. Then, the cross-chain command corresponding to the CCM `module`-`crossChainCommand` combination is verified. If an error occurs either during parameter validation or cross-chain command verification, the channel with the CCM sending chain is terminated and the CCM execution stops.
```

**Review by gkoumout — COMMENTED** [review]
Looks good to me, just left some minor comments. Good job!

**Review by gkoumout — APPROVED** [review]

File: proposals/lip-0053.md [hunk 19]
```diff
@@ -92,7 +92,7 @@ These points guarantee that the CCMs are always forwarded to the correct chains
 The [execution of cross-chain messages][lip-0049#apply] as part of the processing of a cross-chain update is done in several stages, in a way analogous to the execution of a transaction as part of the [processing of a block][lip-0055#block-processing-stages]:
 
 1. **Cross-chain message verification**: Each module can define protocol logic that verifies a CCM, possibly by accessing the state store. If an error occurs, the channel with the CCU sending chain is terminated and the CCU execution stops.
-2. **Cross-chain command verification**: The cross-chain command corresponding to the CCM `module`-`crossChainCommand` combination is verified. If an error occurs, the channel with the CCM sending chain is terminated and the CCM execution stops.
+2. **Cross-chain command verification**: The parameters are decoded using `decode` function and validated against the schema using `validateObjectSchema`. Then, the cross-chain command corresponding to the CCM `module`-`crossChainCommand` combination is verified. If an error occurs either during parameter validation or cross-chain command verification, the channel with the CCM sending chain is terminated and the CCM execution stops.
```
Comment (ricott1) [inline proposals/lip-0053.md:95]: If we finally remove the checks as I suggested, we should revert these changes as well

Comment (janhack) [reply proposals/lip-0053.md:95]: Note that `validateFormat` only validates the root schema so it is still necessary in the `apply` function to validate the CCM parameters.

**Review by sergeyshemyakov — APPROVED** [review]

**Review by janhack — APPROVED** [review]
Looks good.
